watchers/configmap_watcher.go

```diff
@@ -3,44 +3,47 @@ package watchers
 import (
 	"context"
 	"os"
-	"time"
 
 	"github.com/go-logr/logr"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/watch"
-	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/util/retry"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	clientwatch "k8s.io/client-go/tools/watch"
 )
 
 var (
-	logger  logr.Logger
-	backoff = wait.Backoff{
-		Steps:    10,
-		Duration: 10 * time.Second,
-		Factor:   2.0,
-		Jitter:   0.1,
-		Cap:      2 * time.Hour,
-	}
+	logger logr.Logger
 )
 
-func WatchForChanges(clientset kubernetes.Interface, namespace string, configMapName string, dataField string, dataValue string, setupLogger logr.Logger) {
+// watcher implements config.Watch interface which will create a ConfigMapWatcher.
+type watcher struct {
+	configMapGetter v1.ConfigMapInterface
+	configMapName   string
+}
+
+//go:generate mockgen -package=watchers -destination=mock_configmap.go k8s.io/client-go/kubernetes/typed/core/v1 ConfigMapInterface
+//go:generate mockgen -package=watchers -destination=mock_watcher.go k8s.io/apimachinery/pkg/watch Interface
+func WatchForChanges(cmi v1.ConfigMapInterface, configMapName string, dataField string, dataValue string, setupLogger logr.Logger, lastResourceVersion string, exitFunc func()) {
 	logger = setupLogger
-	logger.V(1).Info("watch for changes", "namespace", namespace, "configMap name", configMapName, "data field", dataField, "current value", dataValue)
-	for {
-		var watcher watch.Interface
-		err := createWatcher(clientset, namespace, configMapName, &watcher)
-		if err != nil {
-			logger.Error(err, "cannot create watcher", "namespace", namespace, "configMap name", configMapName)
-			os.Exit(1)
-		}
+	logger.V(1).Info("watch for changes", "configMap name", configMapName, "data field", dataField, "current value", dataValue)
 
-		checkConfigMapChanges(watcher.ResultChan(), dataField, dataValue)
+	wc := watcher{
+		configMapGetter: cmi,
+		configMapName:   configMapName,
 	}
+
+	w, err := clientwatch.NewRetryWatcher(lastResourceVersion, wc)
+	if err != nil {
+		logger.Error(err, "cannot create watcher", "configMap name", configMapName)
+		os.Exit(1)
```

Comment:
shouldn't we use here the exitfunc()?

Reply:
and maybe this should be also tested.

Reply:
do you test every `os.Exit(1)` in the project? especially in `main`?

Reply:
what do you mean not using at all? I'm using it.
Ok. whatever..

Reply:
Actually, I realized that using exitFunc at this point is not safe. If, for some reason, the `exitFunc` does not contain a syscall to exit the flow will continue but with a `nil` watcher which will cause panic. 
I'm pushing the changes but I'm not really ok with it.

Reply:
I'm not sure if I follow you here, the `exitfunc` should work also in all other calls, so it's no longer safe to use?

Reply:
What I wanted to say is that we leave the door open for a panic. If for some reason, the `exitFunc` passed to `WatchForChanges` does not contain an exit call and the `retyWatcher` fails to create the watcher then you got a panic.
But maybe it is just a theoretical case, I admit